### ISSUE TITLE
Go build is slow in Dagger

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#4567]
## Problem

It is commonly reported that building a go program is slower in Dagger. There is no fundamental reason that this should be the case.

## Solution

Help users identify the root cause of their slow go build, and help them fix it. If necessary, fix the root cause in the product and docs.

The most likely cause is the lack of a properly setup cache volume. As a result, each go build runs on an empty go cache, forcing it to scan and re-download all dependencies.

**Comment by nipuna-perera** [comment]
Thank you for opening this! I've been meaning to post in Discord but got sidetracked and when I heard that someone else also suspects their go builds are slow with the SDK I thought I'd bring it up. What I am doing is a multi-arch multi platform build for a go CLI. I will try my best to share as much code as I can while hiding the corporate bits. 

```Go
package main

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"time"

	"dagger.io/dagger"
	platformFormat "github.com/containerd/containerd/platforms"
	"golang.org/x/sync/errgroup"
)

var (
	publishRepo    = "mycompany.docker.repo.com"
	dockerRepoEdge = "mycompany-edge.docker.repo.com"
	platforms      = []dagger.Platform{
		"linux/amd64",
		// "linux/arm64",
	}
	buildImage   = dockerRepoEdge + "/jenkinscore/ubn20/golang-node16:2.4.0-0"
	publishImage = publishRepo + "/myimagename:latest"
)

func main() {
	if err := os.RemoveAll("dist/"); err != nil {
		fmt.Println(err)
	}

	ctx := context.Background()

	client, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stdout))
	if err != nil {
		fmt.Println(err)
	}
	defer client.Close()

	err, variants := build(ctx, client.Pipeline("Build"))
	if err != nil {
		fmt.Println(err)
	}

	if err := publish(ctx, client.Pipeline("Publish"), variants); err != nil {
		fmt.Println(err)
	}

}

// util that returns the architecture of the provided platform
func architectureOf(platform dagger.Platform) string {
	return platformFormat.MustParse(string(platform)).Architecture
}

// util that returns the OS of the provided platform
func platformOf(platform dagger.Platform) string {
	return platformFormat.MustParse(string(platform)).OS
}
func build(ctx context.Context, client *dagger.Client) (error, []*dagger.Container) {
	fmt.Println("Building with Dagger")

	err := dockerLogin(publishRepo, dockerRepoEdge)
	if err != nil {
		return err, nil
	}

	gocache := client.CacheVolume(time.Now().Weekday().String())

	g, ctx := errgroup.WithContext(ctx)

	platformVariants := make([]*dagger.Container, 0, len(platforms))
	for _, platform := range platforms {

		p := platform
		g.Go(func() error {

			golang := client.Container().
				From(buildImage).
				WithEnvVariable("https_proxy", os.Getenv("https_proxy")).
				WithMountedCache("/cache", gocache).
				WithEnvVariable("GOMODCACHE", "/cache").
				WithEnvVariable("GOOS", platformOf(p)).
				WithEnvVariable("GOARCH", architectureOf(p)).
				WithMountedDirectory("/src", client.Host().Workdir()).
				WithWorkdir("/src")

			golang = golang.
				WithExec([]string{"go", "mod", "download"}).
				WithExec([]string{"make", "build"}).
				WithExec([]string{"make", "test"})

			path := "dist"
			output := golang.Directory(path)

			binaryCtr := client.Container(dagger.ContainerOpts{Platform: p})

			binaryCtr = binaryCtr.WithDirectory("/dist", output).
				WithFile("/bin/app", output.File("app")).
				WithEntrypoint([]string{"./bin/app"})
			platformVariants = append(platformVariants, binaryCtr)
			return nil
		})

	}
	if err := g.Wait(); err != nil {
		return err, nil
	}
	return nil, platformVariants
}

func publish(ctx context.Context, client *dagger.Client, variants []*dagger.Container) error {
	img, err := client.Container().
		Publish(ctx, publishImage, dagger.ContainerPublishOpts{
			PlatformVariants: variants})
	if err != nil {
		return err
	}

	fmt.Println(img)
	return nil
}

func dockerLogin(repos ...string) error {
//.....login to company private docker registry
}
```
Note that adding the `errGroup` didn't really seem to make a difference in total time taken. 

Also note that `make build` and `make test` aren't doing anything fancy. Just running `go build` and `go test`. When I run this locally outside of dagger those commands take only a few seconds (less than 10). But when I run within Dagger it takes almost 2 mins per command. I even tried only keeping 1 platform which essentially makes this a single build+test (no parallel). That didnt' seem to improve things. 

When I run consecutively the cache kicks in and the build is almost instantaneous. But if I make any changes then again it takes all that time.

I also did a dagger-graph on this to see if I can identify where it's taking time. And it shows that the build and test steps are what's taking so long. 

I tried increasing the amount of resources I allocate to my docker desktop but it didn't make a difference. I can allocate a lot because I have a M1 Max.

Maybe I am doing something silly that's making it so slow but I can't see it.

**Comment by morlay** [comment]
qemu issue?

**Comment by sipsma** [comment]
I just added a cache mount for the go build cache (which is separate from the go mod cache) to our CI, which seems to make a significant difference (provided your runner isn't ephemeral of course): https://github.com/dagger/dagger/blob/main/internal/mage/util/util.go#L113

There we put it at `/root/.cache/go-build` but you can double check the value by running `go env` in your image and looking at the value of `GOCACHE`. Or I think (not 100% sure) if you set the `GOCACHE` env var to anything custom the go toolchain will honor that

**Comment by nipuna-perera** [comment]
> qemu issue?
> 
> @nipuna-perera
> 
> is there a `linux/arm64` of your buildImage? (Apple m1 need it to run without qemu)

That's a good point. We don't have a arm64 version of it. I can try with a separate image that is arm64

**Comment by nipuna-perera** [comment]
> I just added a cache mount for the go build cache (which is separate from the go mod cache) to our CI, which seems to make a significant difference (provided your runner isn't ephemeral of course): https://github.com/dagger/dagger/blob/main/internal/mage/util/util.go#L113
> 
> There we put it at `/root/.cache/go-build` but you can double check the value by running `go env` in your image and looking at the value of `GOCACHE`. Or I think (not 100% sure) if you set the `GOCACHE` env var to anything custom the go toolchain will honor that

I will try again with the build cache. However, my builds run fast locally outside of Dagger even without a build cache.

**Comment by sipsma** [comment]
> I will try again with the build cache. However, my builds run fast locally outside of Dagger even without a build cache.

Go builds automatically use that build cache, so it should be used outside of dagger unless you are explicitly disabling it there.

**Comment by nipuna-perera** [comment]
It seems that the mismatch in architecture of my builder image is what's slowing things down. So I guess it's related to qemu. When I pull an image that's arm64, the build goes almost as fast as my local. 

It might help to print a warning when the architecture doesn't match that the performance may be affected.

**Comment by shykes** [comment]
Now that we support multi-platform in the API, I wonder if we could implement in Dagger the cross-compile strategy used by https://github.com/crazy-max/goxx ?

**Comment by morlay** [comment]
> It might help to print a warning when the architecture doesn't match that the performance may be affected.

It is a buildkit feature. 

if you engine always run on same machine
it could be set like
```go
client.Container(dagger.ContainerOpts{Platform: fmt.Sprint("linux/%s", runtime.Arch)}).From(buildImage)
```

But pls don't set it by default. 
It will broke runing with remote engine.

**Comment by morlay** [comment]
> Now that we support multi-platform in the API, I wonder if we could implement in Dagger the cross-compile strategy used by https://github.com/crazy-max/goxx ?


It will be better, is client could return the runner platform info

```
type Runner struct {
     Platform string 
     OS string
     Arch string
} 
```

Then we could use to define `--platform=$BUILDPLATFORM` in dagger.
```go
client.Container(dagger.ContainerOpts{Platform: client.Runner().Platform})
```

```Dockerfile
FROM --platform=$BUILDPLATFORM crazymax/goxx:1.17 AS base
```

**Comment by gerhard** [comment]
What should happen next with this? Were any PRs opened for this? If there is some local code, should a PR be created? cc @grouville

**Comment by grouville** [comment]
There are currently no pull requests (PRs) open for this issue. However, there are two proposals being discussed. The issue at hand involves two separate problems that are interconnected but can be solved independently:

#### 1. Incorrectly configured cache volumes:
According to Solomon's statement, the issue seems to be stemming from _incorrectly configured cache volumes._

The fix for this is relatively straightforward, as we need to ensure that people are aware of two key aspects:
- State-of-the-art implementations.
- The iterative debugging process to address this issue.

Currently, our documentation does not cover these points.

Therefore, the first short-term step towards resolving this issue would be to update the documentation and provide the necessary information. The new TUI will assist in identifying such issues, and a documentation upgrade on the example found [here](https://docs.dagger.io/406009/multiplatform-support/#use-cross-compilation) should suffice.

#### 2. Mismatch in the architecture of the builder image:
The second issue is related to a mismatch in the architecture of the builder image.

When building a Go binary on a different architecture than the runner (e.g., building on Mac M1 and running on a different architecture), virtualization is forced without the users being aware of it.

It appears to be related to the use of qemu, as mentioned: "When I pull an image that's arm64, the build goes almost as fast as my local."
Morlay's proposal suggests adding the following code snippet to address this issue:
```go
type Runner struct {
    Platform string
    OS       string
    Arch     string
}

// If client.Container.From() doesn't have an architecture == client.Runner().Platform(), show warning
client.Container(dagger.ContainerOpts{Platform: client.Runner().Platform}).From("")
```

Implementing this solution would enhance the developer experience (DX) by ensuring that all downloaded images preferably match the runner's architecture, thus avoiding virtualization.

If a mismatch occurs, a warning should be displayed to the user, as troubleshooting this behavior requires knowledge and experience in working with containers.

@gerhard:
- First one can be done async, without consensus. 
- Second one is important too, and requires consensus on the DX part

**Comment by grouville** [comment]
## Update on point 2

It turns out that we already retrieve the architecture of the builder. So, my proposal is unnecessary.

For that, Erik implemented the `client.DefaultPlatform()` API: 
```go
package main

import (
	"context"
	"fmt"
	"os"

	"dagger.io/dagger"
)

func main() {
	ctx := context.Background()
	client, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stdout))
	if err != nil {
		panic(err)
	}
	defer client.Close()

	platform, err := client.DefaultPlatform(ctx)
	if err != nil {
		panic(err)
	}

	// initialize this container with the platform of the builder
	ctr := client.Container(dagger.ContainerOpts{Platform: platform})


	// this alpine image has published versions for each of the
	// platforms above. If it was missing a platform,
	// an error would occur when executing a command below.
	ctr = ctr.From("rocker/shiny-verse")

	// execute `uname -m`, which prints the current CPU architecture
	// being executed as
	stdout, err := ctr.
		WithExec([]string{"uname", "-m"}).
		Stdout(ctx)
	if err != nil {
		panic(err)
	}

	// this should print 3 times, once for each of the architectures
	// being executed on
	fmt.Printf("I'm executing on architecture: %s\n", stdout)
}
```

In this example, it does exactly what we want (aka. failure on mismatch):
- `rocker/shiny-verse` is only `linux/amd64`. When trying to use it, we have this error: 
```shell
panic: input:1: container.from no match for platform in manifest sha256:126430df11658bae8b5a93f425f731c820447ff92458eebddca4137f0f046032: not found
```

It seems to be, however, unnecessary: 
```go
package main

import (
	"context"
	"fmt"
	"os"

	"dagger.io/dagger"
)

func main() {
	ctx := context.Background()
	client, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stdout))
	if err != nil {
		panic(err)
	}
	defer client.Close()

	// initialize this container with the platform of the builder
	ctr := client.Container()
	
	// this alpine image has published versions for each of the
	// platforms above. If it was missing a platform,
	// an error would occur when executing a command below.
	ctr = ctr.From("rocker/shiny-verse")

	// execute `uname -m`, which prints the current CPU architecture
	// being executed as
	stdout, err := ctr.
		WithExec([]string{"uname", "-m"}).
		Stdout(ctx)
	if err != nil {
		panic(err)
	}

	// this should print 3 times, once for each of the architectures
	// being executed on
	fmt.Printf("I'm executing on architecture: %s\n", stdout)
}
```

- This gives the same failure output, and I didn't have to specify the platform, Dagger retrieved it

The only moment this mismatch could happen is when the user iterates on valid platform:
```go
platforms      = []dagger.Platform{
		"linux/amd64",
		// "linux/arm64",
	}
```

Here, if the image doesn't have `"linux/arm64"`, the user could still force using a virtualized container:
```go
package main

import (
	"context"
	"fmt"
	"os"

	"dagger.io/dagger"
)

func main() {
	ctx := context.Background()
	client, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stdout))
	if err != nil {
		panic(err)
	}
	defer client.Close()


	// initialize this container with the platform of the builder
	ctr := client.Container(dagger.ContainerOpts{Platform: "linux/amd64"})

	// this alpine image has published versions for each of the
	// platforms above. If it was missing a platform,
	// an error would occur when executing a command below.
	ctr = ctr.From("rocker/shiny-verse")

	// execute `uname -m`, which prints the current CPU architecture
	// being executed as
	stdout, err := ctr.
		WithExec([]string{"uname", "-m"}).
		Stdout(ctx)
	if err != nil {
		panic(err)
	}

	// this should print 3 times, once for each of the architectures
	// being executed on
	fmt.Printf("I'm executing on architecture: %s\n", stdout)
}
```

It runs on my Mac M1. If I were to build a Go binary on this architecture, it would also be longer.

TLDR: The current platform behavior warns users that the builder platform isn't present in the container image. A user would definitely be aware of the fact that virtualization is being used, as it had to force a specific architecture, different than the builder's one.

Following this: this issue can be closed when:
- https://github.com/dagger/dagger/pull/5139
Gets merged
